File: mpis/list2/main.cpp
#include <iostream>
#include <random>
#include <fstream>
#include <vector>

int main() {
    const int minN = 1000; // initial number of bins
    const int maxN = 100000; // final number of bins
    const int k = 50; // number of experiment repeats

    int B_n, U_n, C_n, D_n;
    int B_nSum, U_nSum, C_nSum, D_nSum;
    std::ofstream meansFile("means.txt");
    std::ofstream valuesFile("values.txt");

    std::mt19937 generator(std::random_device{}());

    for (int currentN = minN; currentN <= maxN; currentN += 1000) {
        std::cout << currentN << '\n';
        B_nSum = 0;
        U_nSum = 0;
        C_nSum = 0;
        D_nSum = 0;

        for (int i = 0; i < k; i++) {
            std::vector<int> bins(currentN, 0);
            std::uniform_int_distribution<> distribution(0, currentN - 1); // [0, n - 1]

            int numberOfBalls = 0;
            bool ballsCollided = false;
            bool oneBallInEachBin = false;
            int emptyBins = currentN;
            int binsWithLessThanTwoBalls = currentN;
            while (true) {
                numberOfBalls++;
                const int binNumber = distribution(generator);
                bins[binNumber]++;

                // B_n
                if (!ballsCollided && bins[binNumber] > 1) {
                    B_n = numberOfBalls;
                    B_nSum += B_n;
                    ballsCollided = true;
                }

                if (bins[binNumber] == 1) {
                    emptyBins--;
                }
                else if (bins[binNumber] == 2) {
                    binsWithLessThanTwoBalls--;
                }

                // U_n
                if (numberOfBalls == currentN) {
                    U_n = emptyBins;
                    U_nSum += U_n;
                }

                if (!oneBallInEachBin) {
                    // C_n
                    if (emptyBins == 0) {
                        C_n = numberOfBalls;
                        C_nSum += C_n;
                        oneBallInEachBin = true;
                    }
                }
                else {
                    // D_n
                    if (binsWithLessThanTwoBalls == 0) {
                        D_n = numberOfBalls;
                        D_nSum += D_n;
                        break;
                    }
                }
            }
            valuesFile << currentN << '\t' << B_n << '\t' << U_n << '\t' << C_n
                << '\t' << D_n << '\t' << (D_n - C_n) << '\n';
        }

        meansFile << currentN << '\t'
            << (static_cast<float>(B_nSum) / k) << '\t'
            << (static_cast<float>(U_nSum) / k) << '\t'
            << (static_cast<float>(C_nSum) / k) << '\t'
            << (static_cast<float>(D_nSum) / k) << '\t'
            << (static_cast<float>(D_nSum - C_nSum) / k) << '\n';
    }

    meansFile.close();
    valuesFile.close();

    return 0;
}

File: mpis/list2/main.py
import matplotlib.pyplot as plt
import numpy as np

if __name__ == "__main__":
    n_arr = [n for n in range(1000, 100001, 1000)] # numbers of bins
    k = 50 # number of experiment repeats

    C_n_values = []
    for current_n in n_arr:
        print(current_n)
        B_n_sum = 0
        U_n_sum = 0
        C_n_sum = 0
        D_n_sum = 0

        b_file = open('B_n_values_.txt', 'a')
        u_file = open('U_n_values_.txt', 'a')
        c_file = open('C_n_values.txt', 'a')
        d_file = open('D_n_values_.txt', 'a')
        dc_file = open('D_C_n_values_.txt', 'a')

        for i in range(k):
            bins = np.zeros(current_n, dtype=int)
            number_of_balls = 0
            temp_C_number_of_balls = 0
            ballsCollided = False
            oneBallInEachBin = False
            empty_bins = current_n
            bins_with_less_than_two_balls = current_n
            while True:
                number_of_balls += 1
                binNumber = np.random.randint(0, current_n) # [0, n - 1]
                bins[binNumber] += 1

                # B_n
                if not ballsCollided and bins[binNumber] > 1:
                    B_n_sum += number_of_balls
                    b_file.write(str(number_of_balls) + '\n')
                    ballsCollided = True

                if bins[binNumber] == 1:
                    empty_bins -= 1
                elif bins[binNumber] == 2:
                    bins_with_less_than_two_balls -= 1

                # U_n
                if number_of_balls == current_n:
                    U_n_sum += empty_bins
                    u_file.write(str(empty_bins) + '\n')

                if not oneBallInEachBin:
                    # C_n
                    if empty_bins == 0:
                        C_n_sum += number_of_balls
                        temp_C_number_of_balls = number_of_balls
                        c_file.write(str(number_of_balls) + '\n')
                        oneBallInEachBin = True
                else:
                    # D_n
                    if bins_with_less_than_two_balls == 0:
                        D_n_sum += number_of_balls
                        d_file.write(str(number_of_balls) + '\n')
                        dc_file.write(str(number_of_balls - temp_C_number_of_balls) + '\n')
                        break

        b_file.write(str(B_n_sum / k) + '\n')
        u_file.write(str(U_n_sum / k) + '\n')
        c_file.write(str(C_n_sum / k) + '\n')
        d_file.write(str(D_n_sum / k) + '\n')
        dc_file.write(str((D_n_sum - C_n_sum) / k) + '\n')

    b_file.close()
    u_file.close()
    c_file.close()
    d_file.close()
    dc_file.close()
